tubemq-client-twins/tubemq-client-go/pool/multiplexed.go

```diff
@@ -0,0 +1,386 @@
+	c.state = Connected
+	c.err = nil
+	go c.reader()
+	go c.writer()
+	return nil
+}
+
+func (c *Connection) writer() {
+	var lastErr error
+	for {
+		select {
+		case <-c.done:
+			return
+		default:
+		}
+		req, err := c.buffer.get()
+		if err != nil {
+			lastErr = err
+			break
+		}
+		if err := c.write(req); err != nil {
+			lastErr = err
+			break
+		}
+	}
+	c.close(lastErr, c.done)
+}
+
+func (c *Connection) send(b []byte) error {
+	if c.state == Closed {
+		return ErrConnClosed
+	}
+
+	select {
+	case c.buffer.buffer <- b:
+		return nil
+	case <-c.mDone:
+		return c.err
+	}
+}
+
+func (c *Connection) remove(id uint32) {
+	c.mu.Lock()
+	delete(c.connections, id)
+	c.mu.Unlock()
+}
+
+func (c *Connection) write(b []byte) error {
+	sent := 0
+	for sent < len(b) {
+		n, err := c.conn.Write(b[sent:])
+		if err != nil {
+			return err
+		}
+		sent += n
+	}
+	return nil
+}
+
+func (c *Connection) reader() {
+	var lastErr error
+	for {
+		select {
+		case <-c.done:
+			return
+		default:
+		}
+		rsp, err := c.framer.Decode()
+		if err != nil {
+			lastErr = err
+			break
+		}
+		serialNo := rsp.GetSerialNo()
+		c.mu.RLock()
+		mc, ok := c.connections[serialNo]
+		c.mu.RUnlock()
+		if !ok {
+			continue
+		}
+		mc.reader.recv <- rsp
+		mc.conn.remove(rsp.GetSerialNo())
+	}
+	c.close(lastErr, c.done)
+}
+
+type Multiplexed struct {
+	connections *sync.Map
+}
+
+func (p *Multiplexed) Get(ctx context.Context, address string, serialNo uint32) (*MultiplexedConnection, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	if v, ok := p.connections.Load(address); ok {
+		if c, ok := v.(*Connection); ok {
+			return c.new(ctx, serialNo)
+		}
+		return nil, ErrAssertConnectionFail
+	}
+
+	c := &Connection{
+		address:     address,
+		connections: make(map[uint32]*MultiplexedConnection),
+		done:        make(chan struct{}),
+		mDone:       make(chan struct{}),
+		state:       Initial,
+	}
+	c.buffer = &writerBuffer{
+		buffer: make(chan []byte, queueSize),
+		done:   c.done,
+	}
+	p.connections.Store(address, c)
+
+	conn, dialOpts, err := dial(ctx, address)
+	c.dialOpts = dialOpts
+	if err != nil {
+		return nil, err
+	}
+	c.framer = codec.New(conn)
+	c.conn = conn
+	c.state = Connected
+	go c.reader()
+	go c.writer()
+	return c.new(ctx, serialNo)
+}
+
+func dial(ctx context.Context, address string) (net.Conn, *DialOptions, error) {
+	var timeout time.Duration
+	t, ok := ctx.Deadline()
+	if ok {
+		timeout = t.Sub(time.Now())
+	}
+	dialOpts := &DialOptions{
+		Network: "tcp",
+		Address: address,
+		Timeout: timeout,
+	}
+	select {
+	case <-ctx.Done():
+		return nil, dialOpts, ctx.Err()
+	default:
+	}
+	conn, err := dialWithTimeout(dialOpts)
+	return conn, dialOpts, err
+}
+
+func dialWithTimeout(opts *DialOptions) (net.Conn, error) {
+	if len(opts.CACertFile) == 0 {
+		return net.DialTimeout(opts.Network, opts.Address, opts.Timeout)
+	}
+
+	tlsConf := &tls.Config{}
+	if opts.CACertFile == "none" { // 不需要检验服务证书
```

Comment:
Comments in English are required

Reply:
I have removed this comment because it is obvious.